### DISCUSSION TITLE
Read messages without connecting?

### DISCUSSION THREAD

**Discussion by drurya96** [simpleble/simpleble#282] [Q&A] [answered]
I'm very unfamiliar with bluetooth and working on a simple project to collect data from a few devices. Other examples I'm seeing don't seem to "connect" to the device, rather they seem to listen for all bluetooth signals and filter only the messages relevant to the task at hand. I think my devices aren't staying connected because they aren't really meant to be permanently connected (I think). Does this library offer a way to simply grab all messages as they come, potentially even in an event-driven way?

**Comment by eriklins** [comment, marked as answer]
Bluetooth LE (BLE) comes with two different modes of operation. One would be the devices (e.g. a sensor) broadcasting some data in the BLE advertisement payload which could be read by any central device (e.g. a PC or smartphone). The other is a BLE connection which would allow a central to read/write GATT characteristics of the peripheral to gather data or write e.g. configuration parameters etc. From the above the first scenario likely applies to your use-case.

SimpleBLE does support both. scan_start() or scan_for() in Adapter.h would scan for peripherals and does expose the advertisement payload data into callback functions linked with set_callback_on_scan_found() or set_callback_on_scan_updated() also in Adapter.h. The "scan" example would be a good starting point and you would need to extend the callback functions set_callback_on_scan_found() and/or set_callback_on_scan_updated() with extracting the payload data (likely manufacturer specific data). By default the example just leverage the device name, mac address and rssi value. Functions simpleble_peripheral_manufacturer_data_count() and simpleble_peripheral_manufacturer_data_get() would do the job for manufacturer specific data whereas simpleble_peripheral_services_count() and simpleble_peripheral_services_get() would do it for advertised service data.

Good reading to get better understanding of BLE advertisments would be [here](https://novelbits.io/bluetooth-low-energy-advertisements-part-1/) and [here](https://novelbits.io/bluetooth-low-energy-advertisements-part-2/).

**Reply by drurya96** [reply]
I still need to thoroughly test my code, but a quick run seems to retrieving data at least. It seems `simpleble_peripheral_manufacturer_data_count()` and `simpleble_peripheral_manufacturer_data_get()` both come from src-c instead of src, but after linking those libraries things appear to work. Is there an intended way to do this just using the `Peripheral` object instead of these methods?

I'll go through this reading while traveling, but for now I am a little confused on the getter method that requires an Index. I was under the impression "an advertisement contains 31 bytes of manufacturer data", and this method returns the expected 31 bytes. But what is the index for? Can an advertisement contain a set of manufacturer data? What am I missing here?

**Reply by eriklins** [reply]
Technically the advert payload is up to 31 bytes of raw data. However, the payload is composed of GAP Advertisement Data types like flags, device name, list of service uuids, tx power lever and can also contain a manufacturer specific data type (see above first link). It can actually contain multiple manufacturer specific data types e.g. one in the advert payload and one in the scan response. SimpleBLE does some parsing of the GAP AD types and exposes those in more convenient structure/format. That's why we have different functions to e.g. get the advertised services, service data, manufacturer specific data etc. and why we need indices.
